File: python/beeai_framework/backend/message.py
# Copyright 2025 © BeeAI a Series of LF Projects, LLC
# SPDX-License-Identifier: Apache-2.0

import enum
import json
from collections.abc import Sequence
from datetime import UTC, datetime
from enum import Enum
from typing import Any, Generic, Literal, Required, Self, TypeAlias, TypeVar, cast

from pydantic import BaseModel, ConfigDict, Field, computed_field
from typing_extensions import TypedDict

from beeai_framework.serialization import Serializable
from beeai_framework.utils.dicts import exclude_none
from beeai_framework.utils.lists import cast_list
from beeai_framework.utils.models import to_any_model, to_model
from beeai_framework.utils.strings import to_json

T = TypeVar("T", bound=BaseModel)
T2 = TypeVar("T2")
MessageMeta = dict[str, Any]


class Role(str, Enum):
    ASSISTANT = "assistant"
    SYSTEM = "system"
    TOOL = "tool"
    USER = "user"

    def __str__(self) -> str:
        return self.value

    @classmethod
    def values(cls) -> set[str]:
        return {value for key, value in vars(cls).items() if not key.startswith("_") and isinstance(value, str)}


class SerializableModel(BaseModel, Serializable[dict[str, Any]], auto_register=False):
    """Base class for Pydantic models that need serialization support.

    This class combines BaseModel with Serializable to provide consistent
    serialization behavior for message content parts.
    """

    def create_snapshot(self) -> dict[str, Any]:
        """Create a snapshot of this model for serialization."""
        return self.model_dump()

    def load_snapshot(self, snapshot: dict[str, Any]) -> None:
        """Load state from a snapshot dictionary."""
        for field_name, value in snapshot.items():
            object.__setattr__(self, field_name, value)
        object.__setattr__(self, "__pydantic_fields_set__", set(snapshot.keys()))


class MessageTextContent(SerializableModel):
    type: Literal["text"] = "text"
    text: str


class MessageImageContentImageUrl(TypedDict, total=False):
    url: Required[str]
    detail: str
    format: str


class MessageImageContent(SerializableModel):
    type: Literal["image_url"] = "image_url"
    image_url: MessageImageContentImageUrl


class MessageFileContent(SerializableModel):
    """File content part (e.g. PDF or other document) for multimodal user messages.

    Flattened shape is supported:
        MessageFileContent(file_id="...", format="application/pdf")
        MessageFileContent(file_data="data:application/pdf;base64,...", format="application/pdf")
    """

    type: Literal["file"] = "file"

    file_id: str | None = Field(None, exclude=True)
    file_data: str | None = Field(None, exclude=True)
    filename: str | None = Field(None, exclude=True)
    format: str | None = Field(None, exclude=True)

    @computed_field  # type: ignore[prop-decorator]
    @property
    def file(self) -> dict[str, Any]:
        return exclude_none(
            {"file_id": self.file_id, "file_data": self.file_data, "filename": self.filename, "format": self.format}
        )

    def model_post_init(self, __context: Any) -> None:
        if not (self.file_id or self.file_data):
            raise ValueError("Either 'file_id' or 'file_data' must be provided for MessageFileContent")

    def create_snapshot(self) -> dict[str, Any]:
        """Create snapshot including excluded fields for proper serialization."""
        return {
            "type": self.type,
            "file_id": self.file_id,
            "file_data": self.file_data,
            "filename": self.filename,
            "format": self.format,
        }


class MessageToolResultContent(SerializableModel):
    type: Literal["tool-result"] = "tool-result"
    result: Any
    tool_name: str
    tool_call_id: str


class MessageToolCallContent(SerializableModel):
    type: Literal["tool-call"] = "tool-call"
    id: str
    tool_name: str
    args: str

    def is_valid(self) -> bool:
        if not self.id or not self.tool_name or not self.args:
            return False

        try:
            json.loads(self.args)
            return True
        except Exception:
            return False


class Message(Serializable[dict[str, Any]], Generic[T]):
    id: str | None
    role: Role | str
    content: list[T]
    meta: MessageMeta

    def __init__(self, content: list[T], meta: MessageMeta | None = None, *, id: str | None = None) -> None:
        self.id = id
        self.content = content
        self.meta = meta or {}
        if not self.meta.get("createdAt"):
            self.meta["createdAt"] = datetime.now(tz=UTC)

    @classmethod
    def from_chunks(cls, chunks: Sequence["Message[T]"]) -> Self:
        instance: Self = cls(content=[])
        for chunk in chunks:
            instance.merge(chunk)
        return instance

    def merge(self, other: "Message[T]") -> None:
        self.meta.update(other.meta)
        self.content.extend(other.content)

    @property
    def text(self) -> str:
        return "".join([x.text for x in self.get_texts()])

    def get_texts(self) -> list[MessageTextContent]:
        return [cont for cont in self.content if isinstance(cont, MessageTextContent)]

    def get_by_type(self, tp: type[T2]) -> list[T2]:
        return [cont for cont in self.content if isinstance(cont, tp)]

    def to_plain(self) -> dict[str, Any]:
        return {
            "role": self.role.value if isinstance(self.role, enum.Enum) else self.role,
            "content": [m.model_dump() for m in self.content],
        }

    def to_json_safe(self) -> Any:
        return self.to_plain()

    def __str__(self) -> str:
        return to_json(self.to_plain(), sort_keys=False)

    def clone(self) -> Self:
        return type(self)([c.model_copy() for c in self.content], self.meta.copy())

    def create_snapshot(self) -> dict[str, Any]:
        """Create a snapshot of this message for serialization."""
        return {
            "id": self.id,
            "meta": dict(self.meta),
            "role": str(self.role),
            # Return content as-is - Serializer will handle encoding each item
            "content": list(self.content),
        }

    def load_snapshot(self, snapshot: dict[str, Any]) -> None:
        """Load state from a snapshot dictionary."""
        self.id = snapshot.get("id")
        self.meta = dict(snapshot.get("meta") or {})
        # Content is already deserialized by the Serializer as proper objects
        self.content = snapshot.get("content", [])


AssistantMessageContent = MessageTextContent | MessageToolCallContent


class AssistantMessage(Message[AssistantMessageContent]):
    role = Role.ASSISTANT

    def __init__(
        self,
        content: list[AssistantMessageContent] | AssistantMessageContent | str,
        meta: MessageMeta | None = None,
        *,
        id: str | None = None,
    ) -> None:
        super().__init__(
            [
                (
                    MessageTextContent(text=c)
                    if isinstance(c, str)
                    else to_any_model([MessageToolCallContent, MessageTextContent], cast(AssistantMessageContent, c))
                )
                for c in cast_list(content)
            ],
            meta,
            id=id,
        )

    def get_tool_calls(self) -> list[MessageToolCallContent]:
        return [cont for cont in self.content if isinstance(cont, MessageToolCallContent)]

    def get_text_messages(self) -> list[MessageTextContent]:
        return [cont for cont in self.content if isinstance(cont, MessageTextContent)]


class ToolMessage(Message[MessageToolResultContent]):
    role = Role.TOOL

    def __init__(
        self,
        content: list[MessageToolResultContent] | MessageToolResultContent | str,
        meta: MessageMeta | None = None,
        *,
        id: str | None = None,
    ) -> None:
        super().__init__(
            [
                (
                    MessageToolResultContent.model_validate(json.loads(c))
                    if isinstance(c, str)
                    else to_model(MessageToolResultContent, cast(MessageToolResultContent, c))
                )
                for c in cast_list(content)
            ],
            meta,
            id=id,
        )

    def get_tool_results(self) -> list[MessageToolResultContent]:
        return list(filter(lambda x: isinstance(x, MessageToolResultContent), self.content))


class SystemMessage(Message[MessageTextContent]):
    role = Role.SYSTEM

    def __init__(
        self,
        content: list[MessageTextContent] | MessageTextContent | str,
        meta: MessageMeta | None = None,
        *,
        id: str | None = None,
    ) -> None:
        super().__init__(
            [
                (
                    MessageTextContent(text=c)
                    if isinstance(c, str)
                    else to_model(MessageTextContent, cast(MessageTextContent, c))
                )
                for c in cast_list(content)
            ],
            meta,
            id=id,
        )

    def to_plain(self) -> dict[str, Any]:
        return {
            "role": self.role.value,
            "content": "\n".join([m.text for m in self.content]),
        }


UserMessageContent = MessageTextContent | MessageImageContent | MessageFileContent


class UserMessage(Message[UserMessageContent]):
    role = Role.USER

    def __init__(
        self,
        content: list[UserMessageContent] | UserMessageContent | str,
        meta: MessageMeta | None = None,
        *,
        id: str | None = None,
    ) -> None:
        super().__init__(
            [
                (
                    MessageTextContent(text=c)
                    if isinstance(c, str)
                    else to_any_model(
                        [MessageImageContent, MessageTextContent, MessageFileContent], cast(UserMessageContent, c)
                    )
                )
                for c in cast_list(content)
            ],
            meta,
            id=id,
        )

    @classmethod
    def from_image(cls, data: MessageImageContentImageUrl | str) -> Self:
        """Factory helper to create a user message containing a single image content part.

        Args:
            data: The image content for the user message, either as a URL or a MessageImageContentImageUrl object.
        """
        image_url = MessageImageContentImageUrl(url=data) if isinstance(data, str) else data
        return cls(MessageImageContent(image_url=image_url))

    @classmethod
    def from_file(
        cls,
        *,
        file_id: str | None = None,
        file_data: str | None = None,
        format: str | None = None,
        filename: str | None = None,
    ) -> Self:
        """Factory helper to create a user message containing a single file content part.

        Provide either file_id (for previously uploaded/reference files) or file_data (data URI / base64 encoded).
        Optionally pass format (e.g. "pdf", "txt", "markdown").
        """
        return cls(
            MessageFileContent(
                file_id=file_id,
                file_data=file_data,
                format=format,
                filename=filename,
            )
        )

    @classmethod
    def from_text(cls, text: str) -> Self:
        """Factory helper to create a user message containing a single text content part.

        Args:
            text: The textual content for the user message.
        """
        return cls(MessageTextContent(text=text))


class CustomMessageContent(SerializableModel):
    model_config = ConfigDict(extra="allow")


class CustomMessage(Message[CustomMessageContent]):
    role: str

    def __init__(
        self,
        role: str,
        content: list[CustomMessageContent] | CustomMessageContent | str,
        meta: MessageMeta | None = None,
        *,
        id: str | None = None,
    ) -> None:
        super().__init__(
            [
                (
                    CustomMessageContent.model_validate(MessageTextContent(text=c).model_dump())
                    if isinstance(c, str)
                    else to_model(CustomMessageContent, cast(CustomMessageContent, c))
                )
                for c in cast_list(content)
            ],
            meta,
            id=id,
        )
        self.role = role
        if not self.role:
            raise ValueError("Role must be specified!")


AnyMessage: TypeAlias = Message[Any]


def dedupe_tool_calls(msg: AssistantMessage) -> None:
    final_tool_calls: dict[str, MessageToolCallContent] = {}
    last_id = ""

    excluded_indexes: set[int] = set[int]()
    for idx, chunk in enumerate(msg.content):
        if not isinstance(chunk, MessageToolCallContent):
            continue

        id = chunk.id or last_id
        if id not in final_tool_calls:
            final_tool_calls[id] = chunk.model_copy()
            msg.content[idx] = final_tool_calls[id]
        else:
            excluded_indexes.add(idx)
            last_tool_call = final_tool_calls[id]
            last_tool_call.args += chunk.args

            if not last_tool_call.tool_name:
                last_tool_call.tool_name = chunk.tool_name

        last_id = id

    for idx in sorted(excluded_indexes, reverse=True):
        msg.content.pop(idx)

File: typescript/src/backend/message.ts
/**
 * Copyright 2025 © BeeAI a Series of LF Projects, LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { Serializable } from "@/internals/serializable.js";
import { shallowCopy } from "@/serializer/utils.js";
import { FilePart, ImagePart, TextPart, ToolCallPart, ToolResultPart } from "ai";
import { z } from "zod";
import { ValueError } from "@/errors.js";
import { hasProp, popProp, safeDefineProperty } from "@/internals/helpers/object.js";
import { Logger } from "@/logger/logger.js";
import { isString } from "remeda";
import { watchArray } from "@/internals/helpers/array.js";

export type MessageRole = "user" | "system" | "tool" | "assistant";
export type MessageContentPart = TextPart | ToolCallPart | ImagePart | FilePart | ToolResultPart;

export interface MessageMeta {
  [key: string]: any;
  createdAt?: Date;
}

export interface MessageInput {
  role: MessageRole;
  text: string; // TODO
  meta?: MessageMeta;
}

function isText(content: MessageContentPart): content is TextPart {
  return content.type === "text";
}
function isImage(content: MessageContentPart): content is ImagePart {
  return content.type === "image";
}
function isFile(content: MessageContentPart): content is FilePart {
  return content.type === "file";
}
function isToolCall(content: MessageContentPart): content is ToolCallPart {
  return content.type === "tool-call";
}
function isToolResult(content: MessageContentPart): content is ToolResultPart {
  return content.type === "tool-result";
}

export abstract class Message<
  T extends MessageContentPart = MessageContentPart,
  R extends string = MessageRole | string,
> extends Serializable {
  public abstract readonly role: R;
  public readonly content: T[];

  constructor(
    content: T | T[] | string,
    public readonly meta: MessageMeta = {},
    public id: string | undefined = undefined,
  ) {
    super();
    if (!meta?.createdAt) {
      meta.createdAt = new Date();
    }
    if (typeof content === "string") {
      this.content = [this.fromString(content)];
    } else {
      this.content = Array.isArray(content) ? content : [content];
    }
    for (const chunk of this.content) {
      this.assertContent(chunk);
    }
    this.content = watchArray(this.content, {
      onAdd: this.assertContent,
    });
  }

  protected assertContent(content: T): asserts content is T {}

  protected abstract fromString(input: string): T;

  static of({ role, text, meta }: MessageInput): Message {
    if (role === "user") {
      return new UserMessage(text, meta);
    } else if (role === "assistant") {
      return new AssistantMessage(text, meta);
    } else if (role === "system") {
      return new SystemMessage(text, meta);
    } else if (role === "tool") {
      return new ToolMessage(text, meta);
    } else {
      return new CustomMessage(role, text, meta);
    }
  }

  get text() {
    return this.getTexts()
      .map((c) => c.text)
      .join("");
  }

  getTexts() {
    return this.content.filter(isText) as TextPart[];
  }

  createSnapshot() {
    return { content: shallowCopy(this.content), meta: shallowCopy(this.meta), role: this.role };
  }

  loadSnapshot(snapshot: ReturnType<typeof this.createSnapshot>) {
    Object.assign(this, snapshot);
  }

  toPlain() {
    return { role: this.role, content: shallowCopy(this.content), id: this.id } as const;
  }

  merge(other: Message<T, R>) {
    this.id = this.id || other.id;
    Object.assign(this, other.meta);
    this.content.push(...other.content);
  }

  static fromChunks<M2 extends Message>(this: new (...args: any[]) => M2, chunks: M2[]): M2 {
    const instance = new this([]);
    chunks.forEach((chunk) => instance.merge(chunk));
    return instance;
  }

  [Symbol.iterator](): Iterator<T> {
    return this.content[Symbol.iterator]();
  }
}

export class AssistantMessage extends Message<TextPart | ToolCallPart | FilePart> {
  public readonly role = "assistant";

  static {
    this.register();
  }

  getToolCalls() {
    return this.content.filter(isToolCall);
  }

  protected fromString(text: string): TextPart {
    return { type: "text", text };
  }

  protected assertContent(
    content: TextPart | ToolCallPart | FilePart,
  ): asserts content is TextPart | ToolCallPart | FilePart {
    if (content.type === "tool-call") {
      const key = "args";
      if (!Object.getOwnPropertyDescriptor(content, key)) {
        const args = popProp(content as any, key);

        if (args !== null && !hasProp(content, "input")) {
          Logger.root.warn(
            `The '${key}' property in the AssistantMessage class is deprecated and will be removed. Use the 'input' property for Tool Call content chunks instead.`,
          );
          content.input = args;
        }
      }

      safeDefineProperty(content, key as keyof typeof content, () => {
        Logger.root.warn(
          `The '${key}' property in the Tool Call is deprecated and will be removed. Use the 'input' property instead.`,
        );
        return content.input;
      });
    }
  }
}

export class ToolMessage extends Message<ToolResultPart> {
  public readonly role = "tool";

  static {
    this.register();
  }

  getToolResults() {
    return this.content.filter(isToolResult);
  }

  protected fromString(text: string): ToolResultPart {
    const schema = z
      .object({
        type: z.literal("tool-result"),
        toolName: z.string(),
        toolCallId: z.string(),
      })
      .and(z.object({ output: z.any() }).or(z.object({ result: z.any() })));

    const { success, data } = schema.safeParse(text);
    if (!success) {
      throw new ValueError(`ToolMessage cannot be created from '${text}'!`);
    }

    return data as ToolResultPart;
  }

  protected assertContent(content: ToolResultPart): asserts content is ToolResultPart {
    const result = popProp(content as any, "result", null);
    const isError = popProp(content as any, "isError", false);

    if (result !== null && !hasProp(content, "output")) {
      Logger.root.warn(
        "The 'result' property in the ToolMessage class is deprecated and will be removed. Use the 'output' property for content chunks instead.",
      );
      if (isString(result)) {
        content.output = { type: isError ? "error-text" : "text", value: result };
      } else {
        content.output = { type: isError ? "error-json" : "json", value: result };
      }
    }

    safeDefineProperty(content, "result" as any, () => {
      Logger.root.warn(
        "The 'result' property of Tool Message is deprecated and will be removed. Use the 'output.value' property instead.",
      );
      return content.output.value as any;
    });
    safeDefineProperty(content, "isError" as any, () => {
      Logger.root.warn(
        "The 'isError' property of Tool Message is deprecated and will be removed. Use the 'output.type' property instead.",
      );
      return content.output.type.includes("error");
    });
  }
}

export class SystemMessage extends Message<TextPart> {
  public readonly role: MessageRole = "system";

  static {
    this.register();
  }

  protected fromString(text: string): TextPart {
    return { type: "text", text };
  }
}

export class UserMessage extends Message<TextPart | ImagePart | FilePart> {
  public readonly role = "user";

  static {
    this.register();
  }

  getImages() {
    return this.content.filter(isImage);
  }

  getFiles() {
    return this.content.filter(isFile);
  }

  protected fromString(text: string): TextPart {
    return { type: "text", text };
  }
}

export const Role = {
  ASSISTANT: "assistant",
  SYSTEM: "system",
  USER: "user",
} as const;

export class CustomMessage extends Message<MessageContentPart, string> {
  public role: string;

  constructor(role: string, content: MessageContentPart | string, meta: MessageMeta = {}) {
    super(content, meta);
    if (!role) {
      throw new ValueError(`Role "${role}" must be specified!`);
    }
    this.role = role;
  }

  protected fromString(input: string): MessageContentPart {
    return { type: "text", text: input };
  }
}
